Godep importer

darkowlzz opened the pull request:
### What does this do / why do we need it?

Adds godep config importer support to init.

### What should your reviewer look out for in this PR?

Importer logic and test cases.

### Which issue(s) does this PR fix?

fixes #577

carolynvs commented on cmd/dep/godep_importer.go:
```diff
@@ -0,0 +1,169 @@
+			if err != nil {
+				return nil, nil, err
+			}
+			manifest.Constraints[pc.Ident.ProjectRoot] = gps.ProjectProperties{Source: pc.Ident.Source, Constraint: pc.Constraint}
```
Does godep allow specifying an alternative source? I didn't see anything handling that in `buildProjectConstraint`. If not, we can skip setting the `Source` field on this line since it will always be `""`.

darkowlzz replied:
I couldn't find any way to set alternative source in godep. There's an open issue for the same https://github.com/tools/godep/issues/366 .

carolynvs commented on cmd/dep/godep_importer.go:
```diff
@@ -0,0 +1,169 @@
+		}
+
+		// Rev must not be empty
+		if pkg.Rev == "" {
```
Let's add a unit test to verify error handling when rev is empty too.

carolynvs commented on cmd/dep/godep_importer.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+type godepJSON struct {
+	Name    string         `json:"ImportPath"`
```
What do you think about removing this field since it's not used for any of the conversion logic?

darkowlzz replied:
yeah, doesn't seem to be useful until we decide to make it mandatory for a valid config file.
Removing :)

carolynvs commented on cmd/dep/testdata/godep/expected_import_output.txt:
```diff
@@ -0,0 +1,6 @@
+Detected godep configuration files...
+Converting from Godeps.json ...
+  Using ^0.8.1 as initial constraint for imported dep github.com/sdboyer/deptest
+  Trying ^0.8.1 (3f4c3be) as initial lock for imported dep github.com/sdboyer/deptest
```
This is just an FYI. In #715 I am working on fixing the version set in the lock, and once that goes in the godep importer should get that fix automatically. I'll make sure to verify that this test case is updated, as we shouldn't be locking to a version with an operator, i.e. it should be `Trying v0.8.1` instead of `Trying ^0.8.1` in the feedback. That was my mistake but should be fixed soon!

darkowlzz replied:
@carolynvs #715 didn't change this?

carolynvs replied:
@darkowlzz You just saved my 🥓 ! I missed a spot in godep's importer. Fixing...

darkowlzz commented:
Removed `Name` field from `godepImporter` and added test for missing `Rev`.

Also, added a few checks:
- Skip duplicate package imports
- Skip sub-package imports by deducing project root from the import path

carolynvs commented on cmd/dep/godep_importer.go:
```diff
@@ -0,0 +1,192 @@
+		}
+
+		// Obtain ProjectRoot. Required for avoiding sub-package imports.
+		ip, err := g.sm.DeduceProjectRoot(pkg.ImportPath)
```
What prompted you to check for sub-package imports? I don't remember hitting this with the glide importer and am wondering if this is a general concern for all importers or is godep specific?

carolynvs replied:
Ah okay after scrolling down to the test, it makes sense. Glide lists unique packages and then lists the subpackages used separately which explains why I didn't have to check for this. 😁 Carry on!

darkowlzz commented:
Seems to be ready 😊

carolynvs commented:
🎉💃🏻🚀

sdboyer commented:
woohoo!! 🎉 🚀 🎊
